server/src/services/media.service.ts

```diff
@@ -209,33 +230,67 @@ export class MediaService extends BaseService {
-      ]);
-
-      return { previewPath, thumbnailPath, thumbhash: outputs[2] };
-    } finally {
-      if (didExtract) {
-        await this.storageRepository.unlink(extractedPath);
+    const processInvalidImages = process.env.IMMICH_PROCESS_INVALID_IMAGES === 'true';
+    const colorspace = this.isSRGB(asset) ? Colorspace.SRGB : image.colorspace;
+    const imageIsWebSupported = mimeTypes.isWebSupportedImage(asset.originalFileName);
+    const imageIsRaw = mimeTypes.isRaw(asset.originalFileName);
+
+    const shouldConvertFullsize = !imageIsWebSupported && image.fullsizePreview;
+    const shouldExtractEmbedded = imageIsRaw && image.extractEmbedded;
+    const decodeOptions: DecodeToBufferOptions = {
+      colorspace,
+      processInvalidImages,
+      size: image.preview.size,
+    };
+
+    let useExtracted = false;
+    let decodeInputPath: string = asset.originalPath;
+    // Converted or extracted image from non-web-supported formats (e.g. RAW)
+    let fullsizePath = StorageCore.getImagePath(asset, AssetPathType.FULLSIZE, image.preview.format);
+    if (shouldConvertFullsize) {
+      // unset size to decode fullsize image
+      decodeOptions.size = undefined;
+    }
+
+    if (shouldExtractEmbedded) {
+      // For RAW files, try extracting embedded preview first
+
+      // Assume extracted image from RAW always in JPEG format, as implied from the `jpgFromRaw` tag name
+      const extractedPath = StorageCore.getImagePath(asset, AssetPathType.FULLSIZE, ImageFormat.JPEG);
+      const didExtract = await this.mediaRepository.extract(asset.originalPath, extractedPath, true);
+      useExtracted = didExtract && (await this.shouldUseExtractedImage(fullsizePath, image.preview.size));
+
+      if (useExtracted) {
+        if (shouldConvertFullsize) {
+          // skip re-encoding and directly use extracted as fullsize preview
+          // as usually the extracted image is already heavily compressed, no point doing lossy conversion again
+          fullsizePath = extractedPath;
+        }
+        // use this as origin of preview and thumbnail
+        decodeInputPath = extractedPath;
+        // clone EXIF to persist orientation and other metadata
+        // this is delayed to reduce I/O overhead as we cannot do it in one go with extraction due to exiftool limitations
+        await this.mediaRepository.cloneExif(asset.originalPath, extractedPath);
       }
     }
+
+    const { info, data } = await this.mediaRepository.decodeImage(decodeInputPath, decodeOptions);
+
+    const thumbnailOptions = { colorspace, processInvalidImages, raw: info };
+    const outputs = await Promise.all([
+      this.mediaRepository.generateThumbhash(data, thumbnailOptions),
+      this.mediaRepository.generateThumbnail(data, { ...image.thumbnail, ...thumbnailOptions }, thumbnailPath),
+      this.mediaRepository.generateThumbnail(data, { ...image.preview, ...thumbnailOptions }, previewPath),
+      shouldConvertFullsize &&
```

Comment:
I'd prefer making this a list that you conditionally append to, then waiting for that list with `Promise.all`.